File: research/experiments/sweep/runs/opt-large-csharp-sonnet45/Optimization/Fminbox.cs
namespace Optimization;

/// <summary>
/// Options for box-constrained optimization.
/// </summary>
public class FminboxOptions : OptimizeOptions
{
    public double[]? Lower { get; set; }
    public double[]? Upper { get; set; }
    public string Method { get; set; } = "l-bfgs";
    public double? Mu0 { get; set; }
    public double MuFactor { get; set; } = 0.001;
    public int OuterIterations { get; set; } = 20;
    public double OuterGradTol { get; set; } = 1e-8;
}

/// <summary>
/// Box-constrained optimization via logarithmic barrier method.
/// </summary>
public static class Fminbox
{
    /// <summary>
    /// Minimize a function subject to box constraints: lower <= x <= upper.
    /// Uses logarithmic barrier method with inner unconstrained optimization.
    /// </summary>
    public static OptimizeResult Minimize(
        Func<double[], double> f,
        double[] x0,
        Func<double[], double[]> grad,
        FminboxOptions? options = null)
    {
        var opts = options ?? new FminboxOptions();
        int n = x0.Length;

        // Set up bounds
        double[] lower = opts.Lower ?? Enumerable.Repeat(double.NegativeInfinity, n).ToArray();
        double[] upper = opts.Upper ?? Enumerable.Repeat(double.PositiveInfinity, n).ToArray();

        // Validate bounds
        for (int i = 0; i < n; i++)
        {
            if (lower[i] >= upper[i])
            {
                return new OptimizeResult
                {
                    X = x0,
                    Fun = double.NaN,
                    Gradient = null,
                    Iterations = 0,
                    FunctionCalls = 0,
                    GradientCalls = 0,
                    Converged = false,
                    Message = "Invalid bounds: lower must be less than upper"
                };
            }
        }

        // Nudge initial point to strict interior
        double[] x = NudgeToInterior(x0, lower, upper);

        // Compute initial mu if not provided
        double mu;
        if (opts.Mu0.HasValue)
        {
            mu = opts.Mu0.Value;
        }
        else
        {
            double[] gf = grad(x);
            double[] gb = BarrierGradient(x, lower, upper);
            double gfNorm = VecOps.NormInf(gf);
            double gbNorm = VecOps.NormInf(gb);
            mu = gbNorm > 0 ? opts.MuFactor * gfNorm / gbNorm : 1.0;
        }

        int totalFunctionCalls = 0;
        int totalGradientCalls = 0;

        for (int outerIter = 0; outerIter < opts.OuterIterations; outerIter++)
        {
            // Create barrier-augmented objective
            Func<double[], double> fAug = (xVal) => f(xVal) + mu * BarrierValue(xVal, lower, upper);
            Func<double[], double[]> gradAug = (xVal) =>
            {
                double[] gf = grad(xVal);
                double[] gb = BarrierGradient(xVal, lower, upper);
                return VecOps.Add(gf, VecOps.Scale(gb, mu));
            };

            // Run inner optimizer
            OptimizeResult innerResult;
            if (opts.Method == "bfgs")
            {
                innerResult = BFGS.Minimize(fAug, x, gradAug, opts);
            }
            else if (opts.Method == "l-bfgs")
            {
                var lbfgsOpts = new LBFGSOptions
                {
                    GradTol = opts.GradTol,
                    StepTol = opts.StepTol,
                    FuncTol = opts.FuncTol,
                    MaxIterations = opts.MaxIterations
                };
                innerResult = LBFGS.Minimize(fAug, x, gradAug, lbfgsOpts);
            }
            else if (opts.Method == "conjugate-gradient")
            {
                var cgOpts = new ConjugateGradientOptions
                {
                    GradTol = opts.GradTol,
                    StepTol = opts.StepTol,
                    FuncTol = opts.FuncTol,
                    MaxIterations = opts.MaxIterations
                };
                innerResult = ConjugateGradient.Minimize(fAug, x, gradAug, cgOpts);
            }
            else
            {
                throw new ArgumentException($"Unknown method: {opts.Method}");
            }

            totalFunctionCalls += innerResult.FunctionCalls;
            totalGradientCalls += innerResult.GradientCalls;

            // Clamp result to strict interior
            x = ClampToInterior(innerResult.X, lower, upper);

            // Check projected gradient norm of original objective
            double[] gOrig = grad(x);
            double projGradNorm = ProjectedGradientNorm(x, gOrig, lower, upper);

            if (projGradNorm < opts.OuterGradTol)
            {
                return new OptimizeResult
                {
                    X = x,
                    Fun = f(x),
                    Gradient = gOrig,
                    Iterations = outerIter + 1,
                    FunctionCalls = totalFunctionCalls,
                    GradientCalls = totalGradientCalls,
                    Converged = true,
                    Message = "Converged: projected gradient norm below tolerance"
                };
            }

            // Reduce barrier multiplier
            mu *= opts.MuFactor;
        }

        return new OptimizeResult
        {
            X = x,
            Fun = f(x),
            Gradient = grad(x),
            Iterations = opts.OuterIterations,
            FunctionCalls = totalFunctionCalls,
            GradientCalls = totalGradientCalls,
            Converged = false,
            Message = "Terminated: maximum outer iterations reached"
        };
    }

    /// <summary>
    /// Compute barrier value: sum(-log(x_i - l_i) - log(u_i - x_i)).
    /// Returns Infinity if x is outside the box.
    /// </summary>
    public static double BarrierValue(double[] x, double[] lower, double[] upper)
    {
        double value = 0;
        for (int i = 0; i < x.Length; i++)
        {
            if (!double.IsNegativeInfinity(lower[i]))
            {
                double d = x[i] - lower[i];
                if (d <= 0) return double.PositiveInfinity;
                value -= Math.Log(d);
            }
            if (!double.IsPositiveInfinity(upper[i]))
            {
                double d = upper[i] - x[i];
                if (d <= 0) return double.PositiveInfinity;
                value -= Math.Log(d);
            }
        }
        return value;
    }

    /// <summary>
    /// Compute barrier gradient: component i is -1/(x_i - l_i) + 1/(u_i - x_i).
    /// </summary>
    public static double[] BarrierGradient(double[] x, double[] lower, double[] upper)
    {
        double[] grad = new double[x.Length];
        for (int i = 0; i < x.Length; i++)
        {
            grad[i] = 0;
            if (!double.IsNegativeInfinity(lower[i]))
            {
                grad[i] -= 1.0 / (x[i] - lower[i]);
            }
            if (!double.IsPositiveInfinity(upper[i]))
            {
                grad[i] += 1.0 / (upper[i] - x[i]);
            }
        }
        return grad;
    }

    /// <summary>
    /// Compute projected gradient norm: infinity norm of x - clamp(x - g, lower, upper).
    /// </summary>
    public static double ProjectedGradientNorm(double[] x, double[] g, double[] lower, double[] upper)
    {
        double maxVal = 0;
        for (int i = 0; i < x.Length; i++)
        {
            double projected = x[i] - g[i];
            projected = Math.Max(lower[i], Math.Min(projected, upper[i]));
            double diff = Math.Abs(x[i] - projected);
            if (diff > maxVal) maxVal = diff;
        }
        return maxVal;
    }

    private static double[] NudgeToInterior(double[] x, double[] lower, double[] upper)
    {
        double[] result = VecOps.Clone(x);
        for (int i = 0; i < x.Length; i++)
        {
            if (result[i] <= lower[i])
            {
                if (double.IsPositiveInfinity(upper[i]))
                {
                    result[i] = lower[i] + 1.0;
                }
                else
                {
                    result[i] = 0.99 * lower[i] + 0.01 * upper[i];
                }
            }
            else if (result[i] >= upper[i])
            {
                if (double.IsNegativeInfinity(lower[i]))
                {
                    result[i] = upper[i] - 1.0;
                }
                else
                {
                    result[i] = 0.01 * lower[i] + 0.99 * upper[i];
                }
            }
        }
        return result;
    }

    private static double[] ClampToInterior(double[] x, double[] lower, double[] upper)
    {
        double[] result = new double[x.Length];
        for (int i = 0; i < x.Length; i++)
        {
            double range = upper[i] - lower[i];
            double epsilon = double.IsInfinity(range) ? 1e-8 : 1e-8 * range;
            result[i] = Math.Max(lower[i] + epsilon, Math.Min(x[i], upper[i] - epsilon));
        }
        return result;
    }
}

File: research/experiments/sweep/runs/opt-large-go-sonnet45/fminbox.go
package optimization

import "math"

// FminboxOptions contains box-constrained optimization parameters
type FminboxOptions struct {
	OptimizeOptions
	Lower            []float64
	Upper            []float64
	Method           string  // "bfgs", "l-bfgs", "conjugate-gradient"
	Mu0              *float64 // nil = auto
	MuFactor         float64
	OuterIterations  int
	OuterGradTol     float64
}

// DefaultFminboxOptions returns default fminbox options
func DefaultFminboxOptions(n int, opts *FminboxOptions) FminboxOptions {
	lower := make([]float64, n)
	upper := make([]float64, n)
	for i := range lower {
		lower[i] = math.Inf(-1)
		upper[i] = math.Inf(1)
	}

	options := FminboxOptions{
		OptimizeOptions: DefaultOptions(nil),
		Lower:           lower,
		Upper:           upper,
		Method:          "l-bfgs",
		Mu0:             nil,
		MuFactor:        0.001,
		OuterIterations: 20,
		OuterGradTol:    1e-8,
	}

	if opts != nil {
		if opts.GradTol != 0 {
			options.GradTol = opts.GradTol
		}
		if opts.StepTol != 0 {
			options.StepTol = opts.StepTol
		}
		if opts.FuncTol != 0 {
			options.FuncTol = opts.FuncTol
		}
		if opts.MaxIterations != 0 {
			options.MaxIterations = opts.MaxIterations
		}
		if opts.Lower != nil {
			options.Lower = opts.Lower
		}
		if opts.Upper != nil {
			options.Upper = opts.Upper
		}
		if opts.Method != "" {
			options.Method = opts.Method
		}
		if opts.Mu0 != nil {
			options.Mu0 = opts.Mu0
		}
		if opts.MuFactor != 0 {
			options.MuFactor = opts.MuFactor
		}
		if opts.OuterIterations != 0 {
			options.OuterIterations = opts.OuterIterations
		}
		if opts.OuterGradTol != 0 {
			options.OuterGradTol = opts.OuterGradTol
		}
	}

	return options
}

// BarrierValue computes the logarithmic barrier function value
func BarrierValue(x, lower, upper []float64) float64 {
	sum := 0.0
	for i := range x {
		if !math.IsInf(lower[i], 0) {
			dx := x[i] - lower[i]
			if dx <= 0 {
				return math.Inf(1)
			}
			sum -= math.Log(dx)
		}
		if !math.IsInf(upper[i], 0) {
			dx := upper[i] - x[i]
			if dx <= 0 {
				return math.Inf(1)
			}
			sum -= math.Log(dx)
		}
	}
	return sum
}

// BarrierGradient computes the logarithmic barrier gradient
func BarrierGradient(x, lower, upper []float64) []float64 {
	grad := make([]float64, len(x))
	for i := range x {
		if !math.IsInf(lower[i], 0) {
			grad[i] -= 1.0 / (x[i] - lower[i])
		}
		if !math.IsInf(upper[i], 0) {
			grad[i] += 1.0 / (upper[i] - x[i])
		}
	}
	return grad
}

// ProjectedGradientNorm computes the infinity norm of the projected gradient
func ProjectedGradientNorm(x, g, lower, upper []float64) float64 {
	maxVal := 0.0
	for i := range x {
		// Projected point: clamp(x - g, lower, upper)
		proj := x[i] - g[i]
		proj = math.Max(lower[i], math.Min(upper[i], proj))

		// Projected gradient component
		diff := math.Abs(x[i] - proj)
		if diff > maxVal {
			maxVal = diff
		}
	}
	return maxVal
}

// nudgeToInterior adjusts x to be strictly inside bounds
func nudgeToInterior(x, lower, upper []float64) []float64 {
	xNew := Clone(x)
	for i := range xNew {
		lowerFinite := !math.IsInf(lower[i], 0)
		upperFinite := !math.IsInf(upper[i], 0)

		if lowerFinite && upperFinite {
			// Both bounds finite
			if xNew[i] <= lower[i] || xNew[i] >= upper[i] {
				xNew[i] = 0.99*lower[i] + 0.01*upper[i]
				if xNew[i] <= lower[i] {
					xNew[i] = lower[i] + 1e-8
				}
			}
		} else if lowerFinite && xNew[i] <= lower[i] {
			// Only lower bound finite
			xNew[i] = lower[i] + 1.0
		} else if upperFinite && xNew[i] >= upper[i] {
			// Only upper bound finite
			xNew[i] = upper[i] - 1.0
		}
	}
	return xNew
}

// clampToInterior ensures x stays strictly interior with safety margin
func clampToInterior(x, lower, upper []float64) []float64 {
	xNew := Clone(x)
	for i := range xNew {
		if !math.IsInf(lower[i], 0) {
			xNew[i] = math.Max(lower[i]+1e-15, xNew[i])
		}
		if !math.IsInf(upper[i], 0) {
			xNew[i] = math.Min(upper[i]-1e-15, xNew[i])
		}
	}
	return xNew
}

// Fminbox performs box-constrained optimization using log-barrier method
func Fminbox(
	f func([]float64) float64,
	x0 []float64,
	grad func([]float64) []float64,
	opts *FminboxOptions,
) OptimizeResult {
	n := len(x0)
	options := DefaultFminboxOptions(n, opts)

	// Validate bounds
	for i := range options.Lower {
		if options.Lower[i] >= options.Upper[i] {
			return OptimizeResult{
				X:             x0,
				Fun:           math.Inf(1),
				Gradient:      nil,
				Iterations:    0,
				FunctionCalls: 0,
				GradientCalls: 0,
				Converged:     false,
				Message:       "Invalid bounds: lower >= upper",
			}
		}
	}

	// Nudge x0 to strict interior
	x := nudgeToInterior(x0, options.Lower, options.Upper)

	// Compute initial mu if not provided
	mu := 0.0
	if options.Mu0 != nil {
		mu = *options.Mu0
	} else {
		// Auto-compute from gradient ratio
		gf := grad(x)
		gb := BarrierGradient(x, options.Lower, options.Upper)

		normGf := 0.0
		normGb := 0.0
		for i := range gf {
			normGf += math.Abs(gf[i])
			normGb += math.Abs(gb[i])
		}
		if normGb > 1e-30 {
			mu = options.MuFactor * normGf / normGb
		} else {
			mu = options.MuFactor
		}
	}

	totalFunctionCalls := 0
	totalGradientCalls := 0

	for outerIter := 0; outerIter < options.OuterIterations; outerIter++ {
		// Create barrier-augmented objective
		fAug := func(xVal []float64) float64 {
			fVal := f(xVal)
			bVal := BarrierValue(xVal, options.Lower, options.Upper)
			return fVal + mu*bVal
		}

		gradAug := func(xVal []float64) []float64 {
			gf := grad(xVal)
			gb := BarrierGradient(xVal, options.Lower, options.Upper)
			gAug := make([]float64, len(gf))
			for i := range gAug {
				gAug[i] = gf[i] + mu*gb[i]
			}
			return gAug
		}

		// Run inner optimizer
		var result OptimizeResult
		switch options.Method {
		case "bfgs":
			result = BFGS(fAug, x, gradAug, &options.OptimizeOptions)
		case "l-bfgs":
			lbfgsOpts := &LBFGSOptions{
				OptimizeOptions: options.OptimizeOptions,
				Memory:          10,
			}
			result = LBFGS(fAug, x, gradAug, lbfgsOpts)
		case "conjugate-gradient":
			cgOpts := &ConjugateGradientOptions{
				OptimizeOptions: options.OptimizeOptions,
				Eta:             0.4,
				RestartInterval: n,
			}
			result = ConjugateGradient(fAug, x, gradAug, cgOpts)
		default:
			return OptimizeResult{
				X:             x0,
				Fun:           math.Inf(1),
				Gradient:      nil,
				Iterations:    0,
				FunctionCalls: 0,
				GradientCalls: 0,
				Converged:     false,
				Message:       "Unknown method: " + options.Method,
			}
		}

		totalFunctionCalls += result.FunctionCalls
		totalGradientCalls += result.GradientCalls

		// Clamp result to interior
		x = clampToInterior(result.X, options.Lower, options.Upper)

		// Check projected gradient norm of original objective
		gf := grad(x)
		pgNorm := ProjectedGradientNorm(x, gf, options.Lower, options.Upper)

		if pgNorm < options.OuterGradTol {
			return OptimizeResult{
				X:             x,
				Fun:           f(x),
				Gradient:      gf,
				Iterations:    outerIter + 1,
				FunctionCalls: totalFunctionCalls + 1,
				GradientCalls: totalGradientCalls + 1,
				Converged:     true,
				Message:       "Converged: projected gradient norm below tolerance",
			}
		}

		// Reduce mu
		mu *= options.MuFactor
	}

	// Max outer iterations reached
	gf := grad(x)
	return OptimizeResult{
		X:             x,
		Fun:           f(x),
		Gradient:      gf,
		Iterations:    options.OuterIterations,
		FunctionCalls: totalFunctionCalls + 1,
		GradientCalls: totalGradientCalls + 1,
		Converged:     false,
		Message:       "Maximum outer iterations reached",
	}
}
